### PR TITLE
FUN3D and Funtofem Mesh Morphing Fixes

### PR DESCRIPTION
**Pull request by sean-engelstad**
* Working version of FUN3D mesh morphing
* Working version of Funtofem shape driver for aero shape variables and mesh morphing
* Unit tests provided for these mesh morphing vs. remeshing aero shape tests in `tests/fun3d_tests` folder
* Question: In ESP/CAPS latest CAPStraining example for FUN3D mesh morphing, the `fun3d_aim.unlink()` step occurs right at the end of the forward analysis, but mine has it at the end of the adjoint analysis. Is this a source of error?
* FUN3D mesh morphing derivatives were not as accurate as we would have liked. I noticed that the project_name_body1.dat files which contain the new surface mesh for Fun3dAim mesh morphing only have 7-8 digits of precision. During my finite difference test of lift,drag vs. aoa variable with mesh morphing, only the last ~3 digits are included in $dx_{A0}$. Thus, there appears to be a loss of precision in $x_{A0} + \frac{\partial x_{A0}}{\partial x} dx$ the mesh sensitivity differential for a small change in the design variable. Can we write out these surface mesh files with more precision for the test and/or optimization?

### CODE REVIEW

File: tests/fun3d_tests/fun3d_aero_remesh/run_fun3d_analysis.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 model = FUNtoFEMmodel("wing")
 wing = Body.aeroelastic("wing", boundary=2)
 wing.register_to(model)
-test_scenario = Scenario.steady("turbulent", steps=5000).set_temperature(
+test_scenario = Scenario.steady("euler", steps=5000).set_temperature(
```
Comment (bburke38) [inline tests/fun3d_tests/fun3d_aero_remesh/run_fun3d_analysis.py:31]: Are you sure you want to switch this to Euler?

File: tests/fun3d_tests/fun3d_aero_remesh/test_fun3d_remesh.py [hunk 2]
```diff
@@ -68,7 +67,7 @@ def test_optimizer_file(self):
         ).register_to(wing)
         wing.register_to(model)
         test_scenario = (
-            Scenario.steady("turbulent", steps=10)
+            Scenario.steady("euler", steps=10)
```
Comment (bburke38) [inline tests/fun3d_tests/fun3d_aero_remesh/test_fun3d_remesh.py:70]: Are you sure you want to switch to Euler?

Comment (sean-engelstad) [reply tests/fun3d_tests/fun3d_aero_remesh/test_fun3d_remesh.py:70]: See above comment on mismatching folders.

File: tests/fun3d_tests/fun3d_aero_remesh/run_fun3d_analysis.py (hunk 1 above)
Comment (sean-engelstad) [reply tests/fun3d_tests/fun3d_aero_remesh/run_fun3d_analysis.py:31]: This was already Euler I was just using a mismatching scenario folder called turbulent.

**Comment by sean-engelstad** [comment]
Fixed zero FD error by making sure project name in scenario and fun3d model match the fun3d.nml project name (wasn't matching before). Need to make this a required argument of fun3d_model and auto-register this to scenarios.

**Comment by sean-engelstad** [comment]
Still getting 8% FD error at small step sizes 1e-4, with mesh morphing for both lift and drag. Comparing my results to the ESP/CAPS examples to see if it is due to my code changes or if there is an inherent error in mesh morphing derivatives.

**Comment by sean-engelstad** [comment]
ESP/CAPS mesh morphing example fixes, `data/session08/7_fun3d_OpenMDAO_morph.py`: 'Const' should be 'Cl', "Compressible" should be "compressible". Now our FUN3D fixes for the grid adjoint where `recompute_turb_dist` must be False are crashing the run script. How do I set this in the fun3d.nml auto-generated by ESP/CAPS?

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class TestFuntofemMorph(unittest.TestCase):
+    """
+    This class performs unit test on the oneway-coupled FUN3D driver
+    which uses fixed struct disps or no struct disps
+    TODO : in the case of an unsteady one, add methods for those too?
+    """
+
+    FILENAME = "funtofem-shape-driver.txt"
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:49]:
```suggestion
    FILENAME = "funtofem-nlbgs-aero-driver.txt"
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    FILEPATH = os.path.join(results_folder, FILENAME)
+
+    def test_nominal(self):
+        """test no struct disps into FUN3D"""
+        # build the funtofem model with one body and scenario
+        model = FUNtoFEMmodel("wing")
+        # design the shape
+        # fun3d_model = Fun3dModel.build_morph(csm_file=csm_path, comm=comm, project_name="funtofem_CAPS")
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:57]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    """
+
+    FILENAME = "funtofem-shape-driver.txt"
+    FILEPATH = os.path.join(results_folder, FILENAME)
+
+    def test_nominal(self):
+        """test no struct disps into FUN3D"""
+        # build the funtofem model with one body and scenario
+        model = FUNtoFEMmodel("wing")
+        # design the shape
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:56]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        # fun3d_model = Fun3dModel.build_morph(csm_file=csm_path, comm=comm, project_name="funtofem_CAPS")
+        # aflr_aim = fun3d_model.aflr_aim
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:58]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        # fun3d_aim = fun3d_model.fun3d_aim
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:59]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+        # smaller mesh length is more refined, original value = 5.0
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:61]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        # aflr_aim.set_surface_mesh(ff_growth=1.4, mesh_length=3.0)
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:62]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        # Fun3dBC.inviscid(caps_group="wall").register_to(fun3d_model)
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:63]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+        # farfield = Fun3dBC.Farfield(caps_group="Farfield").register_to(fun3d_model)
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:65]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        # aflr_aim.mesh_sizing(farfield)
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:66]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        # fun3d_model.setup()
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:67]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        # model.flow = fun3d_model
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:68]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aero_morph.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        Variable.shape(name="aoa").set_bounds(
+            lower=-1.0, value=0.0, upper=1.0
+        ).register_to(wing)
+        wing.register_to(model)
+        test_scenario = (
+            Scenario.steady("turbulent", steps=5000)  # 5000
+            .set_temperature(T_ref=300.0, T_inf=300.0)
+            .fun3d_project(fun3d_aim.project_name)
+        )
+        test_scenario.adjoint_steps = 4000  # 2000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_morph.py:149]:
```suggestion
        test_scenario.adjoint_steps = 4000
```

File: tests/fun3d_tests/test_funtofem_aero_morph.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        fun3d_model.setup()
+        model.flow = fun3d_model
+
+        wing = Body.aerothermoelastic("wing", boundary=2)
+        Variable.shape(name="aoa").set_bounds(
+            lower=-1.0, value=0.0, upper=1.0
+        ).register_to(wing)
+        wing.register_to(model)
+        test_scenario = (
+            Scenario.steady("turbulent", steps=5000)  # 5000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_morph.py:145]:
```suggestion
            Scenario.steady("turbulent", steps=5000)
```

File: tests/fun3d_tests/test_funtofem_aero_morph.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+            .set_temperature(T_ref=300.0, T_inf=300.0)
+            .fun3d_project(fun3d_aim.project_name)
+        )
+        test_scenario.adjoint_steps = 4000  # 2000
+        # test_scenario.get_variable("AOA").set_bounds(value=2.0)
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_morph.py:150]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+        Variable.shape(name="aoa").set_bounds(
+            lower=-1.0, value=0.0, upper=1.0
+        ).register_to(wing)
+        wing.register_to(model)
+
+        # add remaining constraints to tacs model
+        caps2tacs.PinConstraint("root").register_to(tacs_model)
+
+        test_scenario = (
+            Scenario.steady("euler", steps=5000)  # 5000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:121]:
```suggestion
            Scenario.steady("euler", steps=5000)
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+            .set_temperature(T_ref=300.0, T_inf=300.0)
+            .fun3d_project(fun3d_aim.project_name)
+        )
+        test_scenario.adjoint_steps = 4000  # 2000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:125]:
```suggestion
        test_scenario.adjoint_steps = 4000 
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+        test_scenario.include(Function.lift()).include(Function.drag())
+        test_scenario.include(Function.ksfailure(ks_weight=10.0))
+        test_scenario.register_to(model)
+
+        # build the solvers and coupled driver
+        solvers = SolverManager(comm)
+        solvers.flow = Fun3dInterface(
+            comm, model, fun3d_dir="meshes", auto_coords=False
+        ).set_units(qinf=1e4)
+        # solvers.structural = None  # will be built by the tacs model at runtime
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:136]:
```suggestion
        # solvers.structural will be built by the TACS model at runtime
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+
+        # add remaining constraints to tacs model
+        caps2tacs.PinConstraint("root").register_to(tacs_model)
+        # 0 means gauge temp in TACS, so matches 300 K ref temp from above
+        caps2tacs.TemperatureConstraint("midplane", temperature=0).register_to(
+            tacs_model
+        )
+
+        test_scenario = (
+            Scenario.steady("turbulent", steps=5000)  # 5000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:231]:
```suggestion
            Scenario.steady("turbulent", steps=5000)
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+            .set_temperature(T_ref=300.0, T_inf=300.0)
+            .fun3d_project(fun3d_aim.project_name)
+        )
+        test_scenario.adjoint_steps = 4000  # 2000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:235]:
```suggestion
        test_scenario.adjoint_steps = 4000
```

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+        test_scenario.include(Function.lift()).include(Function.drag())
+        test_scenario.include(Function.ksfailure(ks_weight=10.0))
+        test_scenario.register_to(model)
+
+        # build the solvers and coupled driver
+        solvers = SolverManager(comm)
+        solvers.flow = Fun3dInterface(
+            comm, model, fun3d_dir="meshes", auto_coords=False
+        ).set_units(qinf=1e4)
+        # solvers.structural = None  # will be built by the tacs model at runtime
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:246]:
```suggestion
        # solvers.structural will be built by the TACS model at runtime
```

File: tests/unit_tests/shape/plots/plot_aero_coord_rel_error.py [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+print(f"\tfirst derivs - 4th order = {first_derivs4}")
+print(f"\tadjoints = {adjoints_vec}")
+print(f"\trel errors - 2nd order = {rel_errors2}")
+print(f"\trel errors - 4th order = {rel_errors4}")
+print(f"\tfunction vals = {function_vec}")
+print(f"\tdf for fp error = {df_vec}")
+print(f"\tdf/f for fp error = {df_ratio_vec}")
+
+
+# plot struct functions, aero functions normalized on the plot to view their third order + nonlinear shapes
```
Comment (bburke38) [inline tests/unit_tests/shape/plots/plot_aero_coord_rel_error.py:117]: Maybe delete these comments in the future.

File: tests/fun3d_tests/test_funtofem_aero_struct_morph.py [hunk 25]
```diff
@@ -0,0 +1,273 @@
...
+
+        # add remaining constraints to tacs model
+        caps2tacs.PinConstraint("root").register_to(tacs_model)
+
+        test_scenario = (
+            Scenario.steady("euler", steps=5000)
+            .set_temperature(T_ref=300.0, T_inf=300.0)
+            .fun3d_project(fun3d_aim.project_name)
+        )
+        test_scenario.adjoint_steps = 4000
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aero_struct_morph.py:125]:
```suggestion
        test_scenario.adjoint_steps = 4000
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+
+    FILENAME = "funtofem-nlbgs-aero-driver.txt"
+    FILEPATH = os.path.join(results_folder, FILENAME)
+
+    def test_nominal(self):
+        """test no struct disps into FUN3D"""
+        # build the funtofem model with one body and scenario
+        model = FUNtoFEMmodel("wing")
+
+
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:57]:
```suggestion
```

File: tests/fun3d_tests/test_funtofem_aeroDV_wing.py [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+
```
Comment (bburke38) [inline tests/fun3d_tests/test_funtofem_aeroDV_wing.py:58]:
```suggestion
```

**Pull request merged**